src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Folding/FoldingRangeEndpoint.cs

```diff
@@ -92,6 +93,57 @@ public TextDocumentIdentifier GetTextDocumentIdentifier(FoldingRangeParams reque
         return foldingRanges;
     }
 
+    private class RazorFileFoldingRangeSyntaxWalker : SyntaxWalker
+    {
+        private readonly RazorSourceDocument _source;
+        internal List<FoldingRange> Ranges { get; }
+
+        public RazorFileFoldingRangeSyntaxWalker(RazorSourceDocument source)
+        {
+            _source = source;
+            Ranges = new List<FoldingRange>();
+        }
+
+        public override void VisitRazorDirective(RazorDirectiveSyntax node)
+        {
+            // first rule: we can fold consecutive usings
+            // if we have an existing range that ends the line *before* this current using directive or begins the line *after*,
+            // we extend the range one line
+            if (node.Body is RazorDirectiveBodySyntax syntax && string.Equals(syntax.Keyword.GetFirstToken().Content, "using"))
+            {
+                var linePosition = node.GetLinePositionSpan(_source);
+
+                var isPartOfExistingRange = false;
+                foreach (var range in Ranges)
```

Comment:
Surely we don't need to look through every range every time we find a using directive? If there is one to continue from, won't it will be the last one in the list?